python/ml4ir/applications/ranking/data/scripts/create_dataset.py

```diff
@@ -0,0 +1,227 @@
+    feature_query_key = feature_config.get_query_key('name')
+    feature_rank = feature_config.get_rank('name')
+    feature_label = feature_config.get_label('name')
+
+    # For ranking data, calculate number of results from data and filter accordingly
+    seed_data, nres = filter_nres(seed_data, feature_query_key, feature_rank, feature_nres, max_num_records, logger)
+
+    # For ranking data, calculate clicked rank from seed data as label_rank
+    seed_data = pd.merge(seed_data,
+                         seed_data[seed_data[feature_label]==True][[feature_rank,feature_query_key]].rename(columns={
+                             feature_rank:'label_rank'}),
+                         on=feature_query_key,
+                         how='left')
+
+    # Now generate data distributions from sample data
+    for mycol in list(seed_data.columns):
+        seed_dict[mycol] = list(seed_data[mycol].values)
+        seed_dict_labeled_pos[mycol] = list(seed_data[seed_data[feature_label] == 1][mycol].values)
+        seed_dict_labeled_neg[mycol] = list(seed_data[seed_data[feature_label] == 0][mycol].values)
+
+    for _ in range(num_samples):
+        # Generate a synthetic query key
+        q_id = generate_key(seed_dict, feature_config.get_query_key('name'))
+
+        # Sample to choose the number of members of the query key (in ranking, the number of results)
+        q_nseq = random.sample(seed_dict[nres], 1)[0]
+
+        # Sample to choose which member is labeled positive (in ranking, which result is clicked),
+        # sampling only from query keys with the same nseq (number of results)
+        q_labelrank = random.sample(list(seed_data[seed_data[nres]==q_nseq].label_rank.values), 1)[0]
+
+        # Build characteristics for each member of the query key (in ranking, each result of the query)
+        for i in range(int(q_nseq)):
+            is_labeled_pos = (i + 1 == q_labelrank)
+            row_dict = {
+                feature_query_key: q_id,
+                nres: q_nseq,
+                'label_rank': q_labelrank,
+                feature_rank: i + 1
+            }
+            for mycol in list(seed_data.columns):
+                # Could also let user specify distribution to sample from for each feature
+                if mycol not in row_dict.keys() and is_labeled_pos:
+                    # row_dict[mycol] = random.sample(seed_dict_labeled_pos[mycol])
+                    row_dict[mycol] = None  # Default value, check type # Could also be functions we pass in
+                elif mycol not in row_dict.keys() and not is_labeled_pos:
+                    # row_dict[mycol] = random.sample(seed_dict_labeled_neg[mycol])
+                    row_dict[mycol] = None
+
+            row_dict = add_feature_highval(row_dict, feature_highval, is_labeled_pos, q_labelrank)
+
+            rows_df.append(row_dict)
+    return pd.DataFrame(rows_df)
+
+
+def main():
+    data_dir = 'ml4ir/applications/ranking/data/example/'
```

Comment:
@calliebradley We should be reading these are command line args, right? Why are we hardcoding these here?

Reply:
Also, the defaults for `data_dir` and `logs_dir` should be the same as `ml4ir.base` as they are under `.gitignore`. We might have a lot of data files checked into github by mistake otherwise. Alternatively you could add `...../data/synthetic/` to the `.gitignore`. But looks like `logs_dir` is not working. My logs were written to current directory.

Reply:
The containment of the script to the one location here was @jakemannix's suggestion as far as I remember? I previously had a separate 'driver' script and I thought it was he who thought that was kind of structural overkill at this point. No way do I want to type them into the command line every time. I can go back to a driver script if we want that instead...

Reply:
@lastmansleeping Also to clarify for the `data_dir`, do you want the default for **both** the seed data and the output synthetic data to be in `python/ml4ir/applications/ranking/data` (that seems dangerous)? Or which one do you want moved? I guess I'd prefer to add these directories to `.gitignore`

Reply:
I remember even I suggested consolidation into a single script. But why does one script mean we can't define arguments? For now, even doing `sys.argv[i]` is fine really. But I'm not a fan of having to change code to run the script. I had to do this for running it on test data.

CC: @jakemannix

Reply:
I want **input** `data_dir` to point to `ml4ir/python/data/` and `logs_dir` to point to `ml4ir/python/logs` to maintain consistency with rest of ml4ir. Of course to use this script in any real way you want to override `data_dir`. 

You can add `ml4ir/python/ml4ir/applications/ranking/data/synthetic` to `.gitignore` and write output data to that directory. That should be fine. Alternatively, you can also create a separate directory for input data(I don't think we should). But make sure you add all data and logs directory defaults to `.gitignore`.

Reply:
"No way do I want to type them into the command line every time" 
I really do not get this. We can define defaults for each arg, right? You would only be specifying the `data_dir` and `feature_config` on each run and those should be mandatory args otherwise the script is not usable.